### PR TITLE
[DAGCombiner][ARM] Teach reduceLoadWidth to handle (and (srl (load), C, ShiftedMask))

### PR DESCRIPTION
**Pull request by topperc**
If we have a shifted mask, we may be able to reduce the load width
to the width of the non-zero part of the mask and use an offset
to the base address to remove the srl. The offset is given by
C+trailingzeros(ShiftedMask).
    
Then we add a final shl to restore the trailing zero bits.
    
I've use the ARM test because that's where the existing (and (srl (load))) tests were.
    
The X86 test was modified to keep the H register.

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -14315,6 +14316,18 @@ SDValue DAGCombiner::reduceLoadWidth(SDNode *N) {
         if ((ExtVT.getScalarSizeInBits() > MaskedVT.getScalarSizeInBits()) &&
             TLI.isLoadExtLegal(ExtType, SRL.getValueType(), MaskedVT))
           ExtVT = MaskedVT;
+      } else if (ExtType == ISD::ZEXTLOAD &&
+                 ShiftMask.isShiftedMask(Offset, ActiveBits) &&
+                 (Offset + ShAmt) < VT.getSizeInBits()) {
+        EVT MaskedVT = EVT::getIntegerVT(*DAG.getContext(), ActiveBits);
+        // If the mask is shifted we can use a narrower load and a shl to insert
+        // the trailing zeros.
+        if (((Offset + ActiveBits) <= ExtVT.getSizeInBits()) &&
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:14325]: ExtVT.getScalarSizeInBits() - not sure if anything other than the isa<ConstantSDNode> ensure this is for scalar only?

Comment (bjope) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:14325]: It is actually guarded on line 14189 (there is an early out for vector loads).

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:14325]: I think I meant to use getScalarSizeInBits(). I'll change it.

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:14325]: `ExtVT.getSizeInBits()` -> `ExtVT.getScalarSizeInBits()` (note this is already used above)

**Review by RKSimon — APPROVED** [review]
LGTM

**Pull request merged**